### PR TITLE
Review

### PR DESCRIPTION
**Pull request by akikoogawa7**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+/runs
+/mlruns
+conda.yaml
+sigmoid_model.py
+.vscode/settings.json
+dataset_preprocessing/all_plant_imgs
+questions.txt
+first_80_idx_plant_labels.csv
+first_80_idx_plant_paths.csv
+combined_plant_labels_and_path.csv
```
Comment (life-efficient) [inline .gitignore:20]: Something must be missing from here... the PR says there are 558 changed files. Unless you've since added an entry to ignore some files and it's talking about deletions.

File: README.md [hunk 2]
```diff
@@ -1 +1,43 @@
...
+- Split data into training/validation set with ratio of 80/20.
+
+#### Feature Extraction
+- CNN class built using Conv2d, BatchNorm2d, LeakyReLU and Dropout. Fully connected networks built using Linear, LeakyReLU and Softmax. 
+- The negative slope for LeakyReLU was set to 0.01.
+#### Regression Analysis
+- Batch runs were logged via `tensorboard`.
+- Accuracy score was used from `torchmetrics`.
+### Outcome
+Highest accuracy score so far for the training set is 80% with `lr = 0.001`, `epochs = 1000`, `kernal_size = 5`. However validation set accuracy is 61% which might mean the model could have overfit the training data too well. This might be due to the small training set which only contains 168 images.<br>
```
Comment (life-efficient) [inline README.md:36]: Present this in tabular form. Add a diagram of the model using [nnsvg](https://alexlenail.me/NN-SVG/)

File: README.md [hunk 3]
```diff
@@ -1 +1,43 @@
-# RHS-Neural-Network
+# RHS Plant Classification Model
+Building a multi-classification model using Convolutional Neural Networks (CNNs)
+
+<img src="/imgs/1.jpg" alt="plant" width="400"/>
+
+### Description
+For this deep learning project, I decided to make use of the images which I had collected from an earlier web scraping project, from the [Royal Horticultural Society database](https://www.rhs.org.uk/Plants/Search-Results?form-mode=true&context=l%3Den%26q%3D%2523all%26sl%3DplantForm). 
+<br><br>
+This CNN model and MVP (Minimum Viable Product) hopes to accurately classify the 3147 different types of plants. However, because of the large amount of possible plant images to be potentially classified, the default number of classes has been finally set to `n_classes = 61` after initially indexing the first 80. Deleting duplicate images and file paths have left us with 61 different plant images to train.<br>
```
Comment (life-efficient) [inline README.md:9]: No need to mention MVP

Comment (life-efficient) [inline README.md:9]: Split this into sections
1. Motivation
2. Preprocessing

File: cnn_model.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+from torch.utils.tensorboard import SummaryWriter
+
+class RHSCNN(torch.nn.Module):
+    def __init__(self, n_classes, in_channels=3, negative_slope=0.01):
+        super().__init__()
+        self.conv_layers = torch.nn.Sequential(
+            torch.nn.Conv2d(in_channels, 32, kernel_size=5),
+            torch.nn.MaxPool2d(5),
+            torch.nn.BatchNorm2d(32),
```
Comment (life-efficient) [inline cnn_model.py:21]: Why only 1 batchnorm? Not a common thing

File: cnn_model.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+from imgs_dataset import RHSImgDataset
+from torch.utils.data.sampler import SubsetRandomSampler
+from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay
+
+from torch.utils.tensorboard import SummaryWriter
+
+class RHSCNN(torch.nn.Module):
+    def __init__(self, n_classes, in_channels=3, negative_slope=0.01):
+        super().__init__()
+        self.conv_layers = torch.nn.Sequential(
```
Comment (life-efficient) [inline cnn_model.py:18]: `self.conv_layers` = misnomer. This is not just the conv layers, it also includes the linear layers.

File: cnn_model.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        )
+
+    def forward(self, x):
+        return self.conv_layers(x)
+
+    def predict(self, x):
+        result = self.forward(x)
+        return torch.argmax(result, dim=1)
+
+def train(model, epochs=1000, lr = 0.001):
```
Comment (life-efficient) [inline cnn_model.py:43]: Remove spaces either side of `=` for lr -_-

File: cnn_model.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+#%%
+predictions = torch.tensor([[1, 1,],[0, 1]])
+labels = class_names
+cm = confusion_matrix(y, y_hat)
+plt.show()
+# %%
+disp = ConfusionMatrixDisplay(confusion_matrix=cm)
+fig, ax = plt.subplots(figsize=(20,20))
+disp.plot(ax=ax)
+plt.show()
```
Comment (life-efficient) [inline cnn_model.py:157]: Put this into a function

File: cnn_model.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+train_features, train_labels = get_train_features_labels()
+val_features, val_labels = get_val_features_labels()
+
+model = RHSCNN(n_classes=n_classes)
+y_hat = model.predict(val_features)
+y = val_labels.detach()
+print(y.shape)
+print()
+print(y_hat.shape)
+
```
Comment (life-efficient) [inline cnn_model.py:147]: Remove

File: cnn_model.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+#%%
+# Load model
+def load_model():
+    PATH = "state_dict_model.pt"
+    CNN.load_state_dict(torch.load(PATH))
+    CNN.eval()
+
+#%%
+class_names = pd.read_csv('first_50_labels.csv')
+class_names
```
Comment (life-efficient) [inline cnn_model.py:136]: Remove

File: cnn_model.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+labels = class_names
+cm = confusion_matrix(y, y_hat)
+plt.show()
+# %%
+disp = ConfusionMatrixDisplay(confusion_matrix=cm)
+fig, ax = plt.subplots(figsize=(20,20))
+disp.plot(ax=ax)
+plt.show()
+
+# %%
```
Comment (life-efficient) [inline cnn_model.py:159]: Add if __name__ == '__main__'

File: combined_plant_labels_and_path.csv [hunk 11]
```diff
@@ -0,0 +1 @@
+,Species,Path
```
Comment (life-efficient) [inline combined_plant_labels_and_path.csv:1]: Stale file?

File: dataset_preprocessing/first_50_labels.csv [hunk 12]
```diff
@@ -0,0 +1,52 @@
+Labels
```
Comment (life-efficient) [inline dataset_preprocessing/first_50_labels.csv:1]: Be careful this first line is not used as a label if someone doesnt know to specify that row 0 is a header

File: dataset_preprocessing/google_img_scraper.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        try:
+            driver.find_element_by_xpath('//*[@id="islmp"]/div/div/div/div/div[5]/input').click()
+            time.sleep(2)
+        except:
+            pass
+        if new_height == last_height:
+            break
+        last_height = new_height
+
+    # Take screenshot of 10 imgs
```
Comment (life-efficient) [inline dataset_preprocessing/google_img_scraper.py:77]: I'm sure you can find the image URLs rather than having to resort to screenshots

File: dataset_preprocessing/preprocess_raw.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+# Preprocess to apply labels to img path
+link_list = []
+def create_url_list():
+    for link in df[0]:
+        link_list.append(link)
+    print(f'Number of url labels: {len(link_list)}')
+
+create_url_list()
+
+label_list = []
```
Comment (life-efficient) [inline dataset_preprocessing/preprocess_raw.py:22]: Put inside function below

File: dataset_preprocessing/preprocess_raw.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+df['Labels'] = df.apply(lambda row: get_url_label(row[0]), axis=1)
+
+def get_first_50_labels():
+    first_50_labels = df['Labels'][0:51]
+    first_50_labels.to_csv('first_50_labels.csv', header=None, index=None)
+
+get_first_50_labels()
+
+# Rename indexed folders with labels
+img_path = os.getcwd() + '/plant_imgs'
```
Comment (life-efficient) [inline dataset_preprocessing/preprocess_raw.py:42]: Use `os.path.join` instead so this works on different operating systems which have different path separators (Windows uses `\` or somethnig)

File: dataset_preprocessing/preprocess_raw.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    print()
+    print(src_path)
+    try:
+        os.rename(src_path, dst_path)
+    except FileNotFoundError as E:
+        print(E)
+    try:
+        os.rename(src_path, dst_path)
+    except OSError as E:
+        print(E)
```
Comment (life-efficient) [inline dataset_preprocessing/preprocess_raw.py:61]: Put all of the actual code run here inside a `if __name__ == '__main__'`. Leave all function definitions outside.

File: dataset_preprocessing/remove_duplicates.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
+from imutils import paths
+import numpy as np
+import argparse
+import cv2
+import os
+
+def dhash(image, hashSize=8):
```
Comment (life-efficient) [inline dataset_preprocessing/remove_duplicates.py:7]: Add docstring and improve function name

File: dataset_preprocessing/remove_duplicates.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+	# grab all image paths with that hash, add the current image path to it, and store the list back in the hashes dictionary
+	p = hashes.get(h, [])
+	p.append(imagePath)
+	hashes[h] = p
+
+# loop over the image hashes
+for (h, hashedPaths) in hashes.items():
+	# check to see if there is more than one image with the same hash
+    if len(hashedPaths) > 1:
+    	# check to see if this is a dry run
```
Comment (life-efficient) [inline dataset_preprocessing/remove_duplicates.py:52]: Please get out of the habit of leaving things out of functions. All of these different small pieces should be bundled into atomic functions, then call them inside a `if __name__ == '__...`

File: img_functions.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+df_to_csv(concatenated_dfs, 'combined_plant_labels_and_path')
+
+# Create img paths list from concatenated df
+img_path_file_list = create_img_paths_list_from_df(concatenated_dfs)
+
+# Drop column from the concatenated df to only keep paths
+paths_df = concatenated_dfs.drop(columns='Species', axis=1)
+df_to_csv_no_header(paths_df, 'first_80_idx_plant_paths')
+
+# 'Paths' column with img paths now no longer have duplicates
```
Comment (life-efficient) [inline img_functions.py:101]: All of this should be put chunked up into functions and then put inside the block below

File: requirements.txt [hunk 20]
```diff
@@ -0,0 +1,7 @@
+torch
+torch.metrics
+tensorboard
+PIL
+os
+urllib.parse
+pandas
```
Comment (life-efficient) [inline requirements.txt:7]: Looks like some things are missing `sklearn`, `multiprocessing` etc

File: sigmoid_model.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+
+# Instantiate model
+model = RHS_NeuralNetwork().float()
+
+# Load in dataloader
+num_workers = 0
+dataloader = torch.utils.data.DataLoader(dataset, batch_size=64, shuffle=True, num_workers=num_workers)
+
+# Train model
+train(model, dataloader)
```
Comment (life-efficient) [inline sigmoid_model.py:97]: This file is called `sigmoid_model` but it seems to also include the training. Move the training into a `main.py` file. Also rename the `sigmoid_model` to `classifier`. Nobody calls it a sigmoid model.

File: sigmoid_model.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+        # Split data
+        X_train, X_test = torch.utils.data.random_split(self.X, [1818, 455])
+        self.X_train = X_train
+
+    def __getitem__(self, index):
+        return self.X[index], self.y[index]
+
+    def __len__(self):
+        return len(self.y)
```
Comment (life-efficient) [inline sigmoid_model.py:41]: THis should be in it's own file

**Review by life-efficient — CHANGES_REQUESTED** [review]
